### PR TITLE
Pickable storage driver clients

### PR DESCRIPTION
**Pull request by antoinejeannot**
Hi !

This PR aims at adding a way to pickle the `ModelLibrary` which, for the moment, is impossible due to storage drivers (especially boto3 and gcs).

It introduces a `MODELKIT_LAZY_DRIVER` environment variable which, if set, will prevent the `Storage Providers` from storing the drivers (boto3, gcs, azure), instead, the configuration settings will be stored allowing the corresponding `Storage Provider` to build it on the fly.

This will make way easier the use of python libraries which use `pickle` such as Apache Spark and multi-processing, leveraging modelkit.

Thanks for reviewing, as always!

### CODE REVIEW

**Review by ldeflandre — APPROVED** [review]
Nicely done.

It would be nice to document the new usage it enables !

**Comment by antoinejeannot** [comment]
@ldeflandre sorry for the missing documentation 🤦
Please take a look at the latest commit!

**Comment by llin17** [comment]
Good job ! I'm wondering the use-case differences between `MODELKIT_LAZT_LOADING` and `MODELKIT_LAZY_DRIVER`. Dose setting `MODELKIT_LAZY_DRIVER=True` cover all the use-case of `MODELKIT_LAZT_LOADING=True` ? If it's the case, should we remove `MODELKIT_LAZT_LOADING` to avoid confusion ?

**Comment by antoinejeannot** [comment]
I would like to wait for @tgenin's approval before merging

**Comment by antoinejeannot** [comment]
@llin17 that's a good remark thanks! Actually the two mechanisms have different goals, and you might want to allow the `MODELKIT_LAZY_DRIVER` to access your cloud provider and disable `MODELKIT_LAZY_LOADING` to build all your models before forking / sending models to workers etc. (with the assumption that your models are pickable of course :))

**Comment by llin17** [comment]
Thanks ! What do you think if we set the default value of `MODELKIT_LAZY_DRIVER` to `True` ?

**Review by ldeflandre — DISMISSED** [review]

File: modelkit/assets/drivers/abc.py [hunk 1]
```diff
@@ -1,10 +1,27 @@
...
+        client_configuration: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        super().__init__()
+        self._client = client
+        self.client_configuration = client_configuration or {}
+        self.bucket = bucket or os.environ.get("MODELKIT_STORAGE_BUCKET") or ""
+        if not self.bucket:
+            raise ValueError("Bucket needs to be set for the storage driver")
+        self.lazy_driver = os.environ.get("MODELKIT_LAZY_DRIVER")
+        if not (self.lazy_driver or client):
```
Comment (tgenin) [inline modelkit/assets/drivers/abc.py:22]: NB: `or self._client`

File: modelkit/assets/drivers/abc.py [hunk 2]
```diff
@@ -34,3 +51,13 @@ def get_object_uri(
         self, object_name: str, sub_part: Optional[str] = None
     ) -> str:  # pragma: no cover
         ...
+
+    @property
+    def client(self):
+        return self._client or self.build_client(self.client_configuration)
+
+    @staticmethod
+    def build_client(
```
Comment (tgenin) [inline modelkit/assets/drivers/abc.py:60]: NB: + `@abstractmethod` ?

File: modelkit/assets/drivers/azure.py [hunk 3]
```diff
@@ -23,18 +23,28 @@ def __init__(
...
-        if client:
-            self.client = client
-        elif connection_string:  # pragma: no cover
-            self.client = BlobServiceClient.from_connection_string(connection_string)
-        else:
-            self.client = BlobServiceClient.from_connection_string(
-                os.environ["AZURE_STORAGE_CONNECTION_STRING"]
+    @staticmethod
+    def build_client(
+        client_configuration: Union[Dict[str, Any], Optional[Dict[str, Any]]]
```
Comment (tgenin) [inline modelkit/assets/drivers/azure.py:40]: NB: just `Optional[Dict[str, Any]]` does not fit ?

**Review by tgenin — APPROVED** [review]

**Comment by tgenin** [comment]
Very nice feature, thank you @antoinejeannot

File: modelkit/assets/drivers/azure.py (hunk 3 above)
Comment (antoinejeannot) [reply modelkit/assets/drivers/azure.py]: I have narrowed it down to a simple `Dict[str, Any]`, thanks for the heads up @tgenin

**Review by ldeflandre — APPROVED** [review]

**Comment by antoinejeannot** [comment]
Thanks for your pretty relevant remarks as always @tgenin, please take a look at the latest commit 🙏

**Review by tgenin — DISMISSED** [review]

**Review by ldeflandre — APPROVED** [review]

**Review by tgenin — APPROVED** [review]

**Pull request merged**